core/src/cost_model.rs

```diff
@@ -0,0 +1,136 @@
+//! `cost_model` aiming to limit the size of broadcasting sets, and reducing the number
+//! of un-parallelizeble transactions (eg, transactions as same writable key sets).
+//! By doing so to improve leader performance.
+
+use solana_sdk::{
+    bpf_loader, bpf_loader_deprecated, bpf_loader_upgradeable, pubkey::Pubkey, system_program,
+};
+use solana_runtime::{
+    hashed_transaction::HashedTransaction,
+};
+use std::{collections::HashMap, str::FromStr};
+
+// TODO  revisit these hardcoded numbers, better get from mainnet log
+const COST_UNIT: u32 = 1;
+const DEFAULT_PROGRAM_COST: u32 = COST_UNIT * 500;
+const CHAIN_MAX_COST: u32 = COST_UNIT * 100_000;
+const BLOCK_MAX_COST: u32 = COST_UNIT * 100_000_000;
+
+use Pubkey as ProgramKey;
+use u32 as Cost;
+
+#[derive(Debug)]
```

Comment:
The code model assigns `cost` to instructions, which is identified by `program id`. I am not sure if this is correct, i am assuming since a transaction can have multiple instructions, the cost_of_transaction = sum( cost_of_instructions ); in order to figure out the cost of transaction, need to iterate through all instructions in it. 

`chain_max_cost` is the configured number of transactions has same `writable accounts`. chained transactions are not parallelizeable. I have question how to define `chained transaction`. I understand transaction have multiple accounts Read-write accounts. Which read-write account should be used to group transactions together as 'chained'?  

`block_max_cost` is the configured total `cost` a block should have. (does using word `block` here cause confusion?)

Reply:
Yes, that's correct about iterating all instructions in the transaction.

All read-write accounts in the transaction should be used to group, since all of those will prevent parallel execution.

block seems reasonable to me.

Reply:
In current version, I 'chain' transactions by signer's account - `&tx.message().account_keys[0..tx.message().header.num_required_signatures as usize];`
What is a right to find out all writable accounts?

Reply:
There is a helper function on Message `is_writable`